// @josh-brown/vector/dist/utilities/aliases.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FloatVector_1 = require("../types/vector/FloatVector");
var FloatMatrix_1 = require("../types/matrix/FloatMatrix");
var vb = FloatVector_1.FloatVector.builder();
var mb = FloatMatrix_1.FloatMatrix.builder();
/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
function vec(data) {
    return vb.fromArray(data);
}
exports.vec = vec;
/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
function mat(data) {
    return mb.fromArray(data);
}
exports.mat = mat;
/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
function eye(size) {
    return mb.identity(size);
}
exports.eye = eye;
function ones(shapeOrDim) {
    return Array.isArray(shapeOrDim) ? mb.ones(shapeOrDim) : vb.ones(shapeOrDim);
}
exports.ones = ones;
function zeros(shapeOrDim) {
    return Array.isArray(shapeOrDim) ? mb.zeros(shapeOrDim) : vb.zeros(shapeOrDim);
}
exports.zeros = zeros;
/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
function diag(elements) {
    return mb.diagonal(vec(elements));
}
exports.diag = diag;
